rclcpp/include/rclcpp/subscription.hpp

```diff
@@ -260,6 +268,13 @@ class Subscription : public SubscriptionBase
     }
     auto typed_message = std::static_pointer_cast<CallbackMessageT>(message);
     any_callback_.dispatch(typed_message, message_info);
+
+    if (subscription_topic_statistics_) {
+      const auto nanos = std::chrono::time_point_cast<std::chrono::nanoseconds>(
+        std::chrono::steady_clock::now());
+      const auto time = rclcpp::Time(nanos.time_since_epoch().count());
```

Comment:
This is a bug, imo. You convert a steady time to a `rclcpp::Time` which default to ROS time, when steady time's epoch and ROS time's epoch are different. Much of the time steady time is counting since the start of the machine, whereas ROS time (and system time) count from the unix epoch on Jan 1, 1970.

Reply:
If you use the `rclcpp::Time` in the signature below you will either need to set the clock time to steady time:

https://github.com/ros2/rclcpp/blob/649d72f835bfca334e7eab39a8c16752094d6317/rclcpp/include/rclcpp/time.hpp#L46
https://github.com/ros2/rcl/blob/b463168528e99024b32c8c2181b96d3cfeb42084/rcl/include/rcl/time.h#L64

Or use ROS/system time and get the current time from the node's clock.

Reply:
I'll set the clock type to steady time. Based on past discussions it was unclear how to obtain the node clock time at the subscriber level.

Reply:
This is another case of interface leaking. Now we've added a dependency on ROS Time from subscriptions, which we need to handle properly. Usually by either adding the clock interface to subscriptions (maybe problematic due to cycles) or by inverting the dependencies and injecting the clock dependency into the subscription some how. This could be done by passing a "get current time" std::function into the subscription which could be a lambda that calls now on the node clock, just as an example. That way the subscription can get the current ros time without requiring that interface.

Reply:
We can continue to use steady time, if that make sense. But if you think system/ros time would make more sense, then there are solutions we just have to make sure we maintain the interfaces correctly.

Reply:
Ok. I think long term time that was obtained from the node's clock makes sense, but we had earlier design discussions (email) where we decided that this was a future improvement given the potential issues with dependency injection. 

As much as I'd like to try and integrate this now I believe we're running out of time (ugh, poor pun) unless you see an easy path forward (I'm currently digging now).

Reply:
No we can do it better in the future, but we should try to follow up. Things like this tend to get lost after a big push like this.